server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -148,13 +148,17 @@ public final class IndicesRequestCache implements RemovalListener<ICacheKey<Indi
     );
 
     /**
-     * If enabled, allows caching all cacheable queries. For now, this means also allowing size > 0 queries.
-     * If enabled, fundamentally non-cacheable queries like DFS queries, queries using the `now` keyword, and
-     * scroll requests are still not cached.
+     * Sets the maximum size of a query which is allowed in the request cache.
+     * This refers to the number of documents returned, not the size in bytes.
+     * Default value of 0 only allows size == 0 queries, matching earlier behavior.
+     * Fundamentally non-cacheable queries like DFS queries, queries using the `now` keyword, and
+     * scroll requests are never cached, regardless of this setting.
      */
-    public static final Setting<Boolean> INDICES_REQUEST_CACHE_ENABLE_FOR_ALL_REQUESTS_SETTING = Setting.boolSetting(
-        "indices.requests.cache.enable_for_all_requests",
-        false,
+    public static final Setting<Integer> INDICES_REQUEST_CACHE_MAX_SIZE_ALLOWED_IN_CACHE_SETTING = Setting.intSetting(
+        "indices.requests.cache.maximum_cacheable_size",
+        0,
+        0,
+        10_000,
```

Comment:
How was this max value determined?

Reply:
Good question. It's honestly a bit arbitrary. I've been testing this with a benchmark based on nyc_taxis, where we run both size = 0 and size > 0 queries in parallel. (They're randomized to make use of the cache). I vary the size for the size > 0 queries, and for each one I compare it to a baseline where size > 0 queries aren't cacheable. I tested sizes 10, 100, 200, 1k, and 5k. 

For 200 and under, allowing size > 0 in the cache significantly improved the latencies for the size > 0 queries, which makes sense, and somewhat improves latencies for the size = 0 queries even though they're competing for the same cache. This is because the cluster is less stressed since it can serve some of the expensive size > 0 queries from the cache, and the size = 200 queries aren't yet so large that they take up a huge amount of the limited cache. 

As size increases to 1k we start seeing each entry taking up significantly more space. We don't see latency improvements for size = 0 anymore since way fewer can fit in the cache. The latency improvements for size > 0 are also less, likely because proportionally more of the request time is spent in the fetch phase, and the request cache doesn't help that. 

By 5k the latency for size = 0 is significantly worse than the baseline and the gain for size > 0 is minimal. So at least for my workload the highest reasonable value is probably close to 1000. But, this could vary a lot based on workload and cache size, so I figure adding some wiggle room for people with different use cases would be reasonable. 

I'm open to changing this max value or removing it entirely, let me know what you think.

Reply:
I think it's probably fine. We already have a limitation on the size of the cache, so this is just a tuning parameter that would allow more to be cached, potentially shortening the TTL of entries, and whether that's good or bad depends on the use case.